README.md
# Icon Animator

Simple and highly configurable icon animation package with icon size, color, duration. It can display any widgets, too.

## Options

- `IconAnimator` can be set with some default options.
- `icon` is the IconData. It can Material icon or any other icons.
- `child` is any widget that can be displayed instead of `icon`.
- If `finish` is set, then it will be displayed when the loop has finished. You may pass `SizedBox.shrink()` to hide the icon.
- `loop` is to set how many loops the `IconAnimator` repeats its children.
- `children` is the frame of animation. Each child is an `AnimationFrame` type with optional setup.
  - If `icon` is set on the animation, then `AnimationFrame` can have color property.
  - `duration` is the duration of each frame.
  - You can use in combination of icon and any widgets.
## Example code

![Icon Animator](https://raw.githubusercontent.com/thruthesky/icon_animator/main/Icon%20Animator.gif)

```dart
IconAnimator(
    icon: Icons.favorite,
    finish: SizedBox.shrink(),
    loop: 3,
    children: [
      AnimationFrame(size: 0, duration: 100),
      AnimationFrame(size: 4, color: Colors.amber[100], duration: 100),
      AnimationFrame(size: 8, color: Colors.amber[200], duration: 100),
      AnimationFrame(size: 16, color: Colors.amber[300], duration: 100),
      AnimationFrame(size: 20, color: Colors.amber[400], duration: 150),
      AnimationFrame(size: 24, color: Colors.amber[500], duration: 200),
      AnimationFrame(size: 28, color: Colors.amber[600], duration: 250),
      AnimationFrame(size: 32, color: Colors.amber[700], duration: 300),
    ],
),
SizedBox(
    width: 32,
    height: 32,
    child: IconAnimator(
    loop: 5,
    finish: SizedBox.shrink(),
    children: [
        AnimationFrame(child: SizedBox.shrink(), duration: 1500),
        AnimationFrame(
            child: Container(
            color: Colors.blue,
            child: Center(
                child: Text('X'),
            ),
            ),
            duration: 300),
        AnimationFrame(
            icon: Icons.thumb_up,
            color: Colors.blue,
            duration: 500),
        AnimationFrame(
            child: Container(
            color: Colors.red,
            child: Center(
                child: Text('O'),
            ),
            ),
            duration: 300),
    ],
    ),
),
```

- A little complicated nested icon animator sample


![Icon Animator 2](https://raw.githubusercontent.com/thruthesky/icon_animator/main/icon_animator_2.gif)

```dart
Positioned(
  child: Center(
    child: IconAnimator(
      loop: 1,
      child: Container(
        padding: EdgeInsets.all(24),
        decoration: BoxDecoration(
            color: Color(0xcc444444),
            borderRadius: BorderRadius.circular(12),
            boxShadow: [
              BoxShadow(
                blurRadius: 8,
                color: Color(0x88555555),
                spreadRadius: 5,
                offset: Offset(1.0, 1.0),
              )
            ]),
        child: Column(
          children: [
            SizedBox(
              width: 84,
              height: 64,
              child: IconAnimator(
                loop: 2,
                children: [
                  AnimationFrame(
                    duration: 250,
                    child: Align(
                      child: SvgIcon(leftSwipe,
                          color: Colors.white, width: 64),
                      alignment: Alignment.centerRight,
                    ),
                  ),
                  AnimationFrame(
                    duration: 250,
                    child: Align(
                      child: SvgIcon(leftSwipe,
                          color: Colors.white, width: 64),
                      alignment: Alignment.centerLeft,
                    ),
                  ),
                ],
              ),
            ),
            SizedBox(height: 32),
            Text(
              '다음 사진을 보시려면,\n왼쪽으로 밀어주세요.',
              style: TextStyle(color: Colors.white),
            )
          ],
        ),
      ),
      finish: SizedBox.shrink(),
      children: [
        AnimationFrame(duration: 1000, child: SizedBox.shrink()),
        AnimationFrame(duration: 5000),
      ],
    ),
  ),
  top: 180,
  left: 0,
  right: 0,
),
```


CHANGELOG.md
## [0.0.4] - mount check

* Add mount check to avoid unmount exception.
  
## [0.0.3] - Document update

* Document update only.
## [0.0.2] - Add finising widget and child option

* When loop has finished, it can show a widget. `SizedBox.shrink()` may be used to hide the icon.
* It can display any widgets instead of icon or combination of icon and widgets.
* ## [0.0.1] - Icon resize animation

* First verion

test/icon_animator_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:icon_animator/icon_animator.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//     expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//   });
// }


lib/icon_animator.dart
library icon_animator;

import 'dart:async';

import 'package:flutter/material.dart';

class AnimationFrame {
  AnimationFrame({
    this.size,
    this.color,
    @required this.duration,
    this.icon,
    this.child,
  });
  final double size;
  final Color color;
  final int duration;
  final IconData icon;
  final Widget child;
}

class IconAnimator extends StatefulWidget {
  IconAnimator({
    this.icon,
    this.finish,
    @required this.children,
    this.loop,
    this.child,
  });
  final IconData icon;
  final Widget finish;
  final int loop;
  final List<AnimationFrame> children;
  final Widget child;
  @override
  _IconAnimatorState createState() => _IconAnimatorState();
}

class _IconAnimatorState extends State<IconAnimator> {
  int frameCount = 0;
  int loopCount = 0;
  AnimationFrame animate;
  @override
  void initState() {
    super.initState();

    resizeLoop();
  }

  resizeLoop() {
    if (!mounted) return;
    setState(() {
      animate = widget.children[frameCount];
    });
    Timer(
      Duration(milliseconds: widget.children[frameCount].duration),
      () {
        frameCount++;
        if (frameCount >= widget.children.length) {
          if (widget.loop != null) {
            loopCount++;
            if (loopCount >= widget.loop) {
              // finished
              if (mounted) setState(() {});
              return;
            }
          }
          frameCount = 0;
        }
        resizeLoop();
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    if (widget.loop != null && loopCount >= widget.loop) {
      if (widget.finish != null) {
        return widget.finish;
      }
    }
    Widget child;
    if (animate.child != null) {
      child = animate.child;
    } else if (widget.child != null) {
      child = widget.child;
    } else if (animate.icon != null) {
      child = Icon(animate.icon);
    } else if (widget.icon != null) {
      child = Icon(widget.icon);
    } else {
      child = SizedBox.shrink();
    }
    if (child is Icon) {
      return Icon(
        animate.icon != null ? animate.icon : widget.icon,
        size: animate.size,
        color: animate.color,
      );
    } else {
      return child;
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:icon_animator/icon_animator.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'Icon Animator',
            ),
            SizedBox(height: 40),
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                SizedBox(
                  width: 32,
                  height: 32,
                  child: IconAnimator(
                    icon: Icons.radio_button_unchecked,
                    loop: 1,
                    children: [
                      AnimationFrame(size: 0, duration: 100),
                      AnimationFrame(size: 4, color: Colors.red, duration: 100),
                      AnimationFrame(size: 8, duration: 100),
                      AnimationFrame(
                          size: 12, color: Colors.red, duration: 100),
                      AnimationFrame(size: 16, duration: 100),
                      AnimationFrame(
                          size: 20, color: Colors.blue, duration: 150),
                      AnimationFrame(
                          size: 24, color: Colors.blue, duration: 200),
                      AnimationFrame(
                          size: 28, color: Colors.blue, duration: 250),
                      AnimationFrame(
                          size: 32, color: Colors.blue, duration: 300),
                    ],
                  ),
                ),
                SizedBox(
                  width: 32,
                  height: 32,
                  child: IconAnimator(
                    icon: Icons.lens,
                    loop: 2,
                    children: [
                      AnimationFrame(size: 0, duration: 100),
                      AnimationFrame(
                          size: 4, color: Colors.green, duration: 100),
                      AnimationFrame(
                          size: 8, color: Colors.green, duration: 100),
                      AnimationFrame(
                          size: 16, color: Colors.green, duration: 100),
                      AnimationFrame(
                          size: 20, color: Colors.green, duration: 150),
                      AnimationFrame(
                          size: 24, color: Colors.green, duration: 200),
                      AnimationFrame(
                          size: 28, color: Colors.green, duration: 250),
                      AnimationFrame(
                          size: 32, color: Colors.green, duration: 300),
                    ],
                  ),
                ),
                SizedBox(
                  width: 32,
                  height: 32,
                  child: IconAnimator(
                    icon: Icons.favorite,
                    finish: SizedBox.shrink(),
                    loop: 3,
                    children: [
                      AnimationFrame(size: 0, duration: 100),
                      AnimationFrame(
                          size: 4, color: Colors.amber[100], duration: 100),
                      AnimationFrame(
                          size: 8, color: Colors.amber[200], duration: 100),
                      AnimationFrame(
                          size: 16, color: Colors.amber[300], duration: 100),
                      AnimationFrame(
                          size: 20, color: Colors.amber[400], duration: 150),
                      AnimationFrame(
                          size: 24, color: Colors.amber[500], duration: 200),
                      AnimationFrame(
                          size: 28, color: Colors.amber[600], duration: 250),
                      AnimationFrame(
                          size: 32, color: Colors.amber[700], duration: 300),
                    ],
                  ),
                ),
                SizedBox(
                  width: 32,
                  height: 32,
                  child: IconAnimator(
                    loop: 5,
                    finish: SizedBox.shrink(),
                    children: [
                      AnimationFrame(child: SizedBox.shrink(), duration: 1500),
                      AnimationFrame(
                          child: Container(
                            color: Colors.blue,
                            child: Center(
                              child: Text('X'),
                            ),
                          ),
                          duration: 300),
                      AnimationFrame(
                          icon: Icons.thumb_up,
                          color: Colors.blue,
                          duration: 500),
                      AnimationFrame(
                          child: Container(
                            color: Colors.red,
                            child: Center(
                              child: Text('O'),
                            ),
                          ),
                          duration: 300),
                    ],
                  ),
                ),
                IconAnimator(
                  icon: null,
                  children: [
                    AnimationFrame(
                        icon: Icons.volume_mute, size: 24, duration: 100),
                    AnimationFrame(
                        icon: Icons.volume_down, size: 24, duration: 100),
                    AnimationFrame(
                        icon: Icons.volume_up, size: 24, duration: 100)
                  ],
                ),
                IconAnimator(
                  icon: null,
                  children: [
                    AnimationFrame(
                        icon: Icons.notifications, size: 24, duration: 100),
                    AnimationFrame(
                        icon: Icons.notifications_active,
                        size: 24,
                        duration: 100),
                  ],
                ),
              ],
            ),
            SizedBox(
              width: 128,
              height: 128,
              child: IconAnimator(
                icon: Icons.gps_fixed,
                children: [
                  AnimationFrame(size: 0, duration: 100),
                  AnimationFrame(
                      size: 4, color: Colors.grey[100], duration: 100),
                  AnimationFrame(
                      size: 8, color: Colors.grey[200], duration: 100),
                  AnimationFrame(
                      size: 16, color: Colors.grey[300], duration: 100),
                  AnimationFrame(
                      size: 20, color: Colors.grey[400], duration: 150),
                  AnimationFrame(
                      size: 24, color: Colors.grey[500], duration: 200),
                  AnimationFrame(
                      size: 28, color: Colors.grey[600], duration: 250),
                  AnimationFrame(
                      size: 32, color: Colors.grey[700], duration: 300),
                  AnimationFrame(
                      size: 38, color: Colors.grey[800], duration: 300),
                  AnimationFrame(
                      size: 56, color: Colors.grey[900], duration: 300),
                  AnimationFrame(
                      size: 82, color: Colors.grey[900], duration: 300),
                  AnimationFrame(
                      size: 128, color: Colors.grey[900], duration: 300),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}
